File: Data Structures & Algorithms/permutations/submission-7.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> current = new ArrayList<>();
        boolean[] isUsed = new boolean[nums.length];

        backtrack(nums, isUsed, current, result);

        return result;
    }

    public void backtrack(int[] nums, boolean[] isUsed, List<Integer> current, List<List<Integer>> result){
        if(current.size() == nums.length){
            result.add(new ArrayList<>(current));
            return;
        }

        for(int i = 0; i < nums.length; i++){
            if(!isUsed[i]){
                isUsed[i] = true;
                current.add(nums[i]);
                backtrack(nums, isUsed, current, result);
                current.remove(current.size() - 1);
                isUsed[i] = false;
            }
        }
    }
}

File: Data Structures & Algorithms/combination-target-sum/submission-7.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        result = []
        self.dfs(0, result, [], nums, 0, target)
        return result
    
    def dfs(self, i, result, current, nums, sum, target):

        if sum == target:
            result.append(current.copy())
            return
        
        if i >= len(nums) or sum > target:
            return
        
        current.append(nums[i])
        self.dfs(i, result, current, nums, sum + nums[i], target)
        current.pop()

        self.dfs(i + 1, result, current, nums, sum, target)






        